### DISCUSSION TITLE
/tmp folder used by frankenphp is not the /tmp folder that I can access via ssh ??

### DISCUSSION THREAD

**Discussion by geoidesic** [php/frankenphp#1060] [Q&A] [answered]
Hi, I'm very confused..

If I upload a PHP script and then have that php script list the contents of the /tmp folder, the contents are completely different to if I ssh into the server and do `ls /tmp`?

I guess that frankenphp has its own tmp folder squirrelled away somewhere that it's using. How can I find the actual path to that?

**Comment by dunglas** [comment]
FrankenPHP doesn't have its own `/tmp` folder. Your issue is likely unrelated to FrankenPHP.

**Reply by geoidesic** [reply]
What else could it be? 

If I run this on the CLI `php -r "echo shell_exec('ls /tmp');"` I get a different result to if I put it in a php file and then hit the php file.

Here's the php file (index.php):
```
<?php
echo '<pre>';
print_r(shell_exec('ls /tmp'));
```

So this is not something wrong with the code. It's just a vanilla php file. I haven't installed any additional software other than frankenphp.

The output from the cli is
```
# php -r "echo shell_exec('ls /tmp');"
CLRTC-upload.csv
sess_04bqdrs9j75qtpa2q44oud2cc0 
... 
```

The output from the browsed php file is:
```
sess_004e07577cb7796e85fabcb7e9a0d7b3
```

If I run that index.php file from the CLI, then it outputs a different list than if I browse to it in the browser.

The CLI php is a separate install of PHP, because frankenphp doesn't provide its own CLI for PHP. 

So I can't see how this could be unrelated to frankenphp, since it only occurs when using frankenphp?

**Reply by withinboredom** [reply]
Your environment likely has the temporary file location customized. Check the output of `tempnam()` to see where it puts the files. Not all temporary files go to /tmp, which is usually reserved for system files, not arbitrary applications (at least on a production system). Some admins will create a ramdisk for temporary files instead of /tmp as well.

**Reply by geoidesic** [reply]
Ok I tried that. `tempnam` requires arguments, can't be run without them, so presumably you meant this:
```
<?php
echo '<pre>';
$result = tempnam("/tmp", "FOO");
print_r($result);
echo "\n";
print_r(sys_get_temp_dir());
echo "\n";
```
The output from running the script via the browser is
```
/tmp/FOOMeNIMn
/tmp
```
However, on the CLI that file is not created in `/tmp`. 
Once again, I can only find it if I browse to a php script that includes `print_r(shell_exec('ls /tmp'));`

I'd be happy to give you root access to this server if you would like to see it in action.

**Reply by geoidesic** [reply]
Interestingly... I decided to change the root folder in the Caddyfile, and that reset the contents of the ghost /tmp folder. So it feels like it is a /tmp folder that is related to the virtual host that's set up in the caddy file.

I went a step further and noticed that the ghost /tmp folder empties each time I restart the frankenphp systemd. Further evidence that it's related to frankenphp or at least to the service.

**Reply by geoidesic** [reply, marked as answer]
Ah... It must be a systemd thing. I notice this setting in systemd/frankenphp:

```
PrivateTmp=true
```

**Reply by withinboredom** [reply]
> Ok I tried that. tempnam requires arguments, can't be run without them, so presumably you meant this:

Per documentation:

>Creates a file with a unique filename, with access permission set to 0600, in the specified directory. If the directory does not exist or is not writable, tempnam() may generate a file in the system's temporary directory, and return the full path to that file, including its name.

So `tempnam('/')` (assuming it isn't running as root), will return the current temporary directory.

In any case, glad you figured it out.
